### PR TITLE
crypto: add BLS12381 validator key type

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Description

This pull request adds bls12381 keys to tendermint. The Dash ecosystem heavily relies on BLS keys in the masternode network for threshold signing and distributed key generation. Out of scope of this pull request is all key or signature aggregation and threshold signing.

### CODE REVIEW

File: abci/example/kvstore/persistent_kvstore.go [hunk 1]
```diff
@@ -239,12 +239,12 @@ func (app *PersistentKVStoreApplication) execValidatorTx(tx []byte) types.Respon
...
 
 	// update
-	return app.updateValidator(types.Ed25519ValidatorUpdate(pubkey, power))
+	return app.updateValidator(types.BLS12381ValidatorUpdate(pubkey, power))
 }
 
 // add, update, or remove a validator
 func (app *PersistentKVStoreApplication) updateValidator(v types.ValidatorUpdate) types.ResponseDeliverTx {
-	key := []byte("val:" + string(v.PubKey.GetEd25519()))
+	key := []byte("val:" + string(v.PubKey.GetBls12381()))
```
Comment (alexanderbez) [inline abci/example/kvstore/persistent_kvstore.go:247]: nit: I'd call this `GetBLS12381`

File: crypto/bls12381/bls12381.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func (pubKey PubKey) VerifyBytes(msg []byte, sig []byte) bool {
+	// make sure we use the same algorithm to sign
+	if len(sig) != SignatureSize {
+		return false
+	}
+	publicKey, err := bls.PublicKeyFromBytes(pubKey)
+	if err != nil {
+		// maybe log/panic?
```
Comment (alexanderbez) [inline crypto/bls12381/bls12381.go:155]: I do not recommend panicking in these methods.

File: crypto/crypto.go [hunk 3]
```diff
@@ -24,15 +34,17 @@ type PubKey interface {
 	Bytes() []byte
 	VerifyBytes(msg []byte, sig []byte) bool
 	Equals(PubKey) bool
-	Type() string
+	TypeIdentifier() string
```
Comment (alexanderbez) [inline crypto/crypto.go:37]: What is the `TypeIdentifier` used for and how does it differ from `Type`?

File: light/helpers_test.go [hunk 4]
```diff
@@ -39,7 +43,7 @@ func genPrivKeys(n int) privKeys {
 
 // Extend adds n more keys (to remove, just take a slice).
 func (pkz privKeys) Extend(n int) privKeys {
-	extra := genPrivKeys(n)
+	extra := genPrivKeys(n,crypto.BLS12381)
```
Comment (alexanderbez) [inline light/helpers_test.go:46]: formatting seems off in this file.

File: light/helpers_test.go [hunk 5]
```diff
@@ -21,10 +21,14 @@ import (
 type privKeys []crypto.PrivKey
 
 // genPrivKeys produces an array of private keys to generate commits.
-func genPrivKeys(n int) privKeys {
+func genPrivKeys(n int, keyType crypto.KeyType) privKeys {
 	res := make(privKeys, n)
 	for i := range res {
-		res[i] = ed25519.GenPrivKey()
+		if keyType == crypto.BLS12381 {
```
Comment (alexanderbez) [inline light/helpers_test.go:27]: I know these are the two supported types, but maybe we want to default to edwards or error when the type doesn't match? Otherwise, we'll get a nil private key.

File: light/verifier_test.go [hunk 6]
```diff
@@ -24,7 +25,7 @@ func TestVerifyAdjacentHeaders(t *testing.T) {
 	)
 
 	var (
-		keys = genPrivKeys(4)
+		keys = genPrivKeys(4,crypto.BLS12381)
```
Comment (alexanderbez) [inline light/verifier_test.go:28]: nit: formatting

File: state/tx_filter.go [hunk 7]
```diff
@@ -10,6 +11,7 @@ import (
 func TxPreCheck(state State) mempl.PreCheckFunc {
 	maxDataBytes := types.MaxDataBytesUnknownEvidence(
 		state.ConsensusParams.Block.MaxBytes,
+		crypto.BLS12381,
```
Comment (alexanderbez) [inline state/tx_filter.go:14]: Why is this change necessary?

File: state/tx_filter_test.go [hunk 8]
```diff
@@ -16,7 +16,7 @@ import (
 
 func TestTxFilter(t *testing.T) {
 	genDoc := randomGenesisDoc()
-	genDoc.ConsensusParams.Block.MaxBytes = 3000
+	genDoc.ConsensusParams.Block.MaxBytes = 3080
```
Comment (alexanderbez) [inline state/tx_filter_test.go:19]: Reasoning for this change?

File: types/block.go [hunk 9]
```diff
@@ -283,12 +283,12 @@ func BlockFromProto(bp *tmproto.Block) (*Block, error) {
 // MaxDataBytes returns the maximum size of block's data.
 //
 // XXX: Panics on negative result.
-func MaxDataBytes(maxBytes int64, valsCount, evidenceCount int) int64 {
+func MaxDataBytes(maxBytes int64, keyType crypto.KeyType, valsCount, evidenceCount int) int64 {
```
Comment (alexanderbez) [inline types/block.go:286]: Why do we need a key type provided in these?

**Review by alexanderbez — COMMENTED** [review]
Changes look great so far. I can't speak to the CGO or actual BLS code, but so far it seems to be on the right path 👍

File: abci/example/kvstore/persistent_kvstore.go (hunk 1 above)
Comment (QuantumExplorer) [reply abci/example/kvstore/persistent_kvstore.go:247]: Sounds good, will change it.

File: crypto/bls12381/bls12381.go (hunk 2 above)
Comment (QuantumExplorer) [reply crypto/bls12381/bls12381.go:155]: Got it

File: crypto/crypto.go (hunk 3 above)
Comment (QuantumExplorer) [reply crypto/crypto.go:37]: I found the word Type for a string wasn't ideal. I really wanted to use Type for an Enum of possible key types. So I renamed Type to TypeIdentifier.

File: light/helpers_test.go (hunk 4 above)
Comment (QuantumExplorer) [reply light/helpers_test.go:46]: Do you mean add a space after the comma?

File: light/helpers_test.go (hunk 5 above)
Comment (QuantumExplorer) [reply light/helpers_test.go:27]: I actually noticed we were generating BLS keys for both key types. Fixed it and also defaulting to edwards.

File: state/tx_filter.go (hunk 7 above)
Comment (QuantumExplorer) [reply state/tx_filter.go:14]: Its because signature sizes are different in BLS and ED keys, which will change the max data bytes.
BLS public keys are 48 bytes and sigs are 96 bytes. Ed public keys are 32 bytes and sigs are 64 bytes. So a 16 byte difference for a public key and a 32 byte difference for the signature.

`maxEvidenceBytes := int64(maxNumEvidence) * MaxEvidenceBytesForKeyType(keyType)
	maxDataBytes := maxBytes -
		MaxOverheadForBlock -
		MaxHeaderBytes -
		int64(valsCount)*MaxVoteBytesForKeyType(keyType) -
		maxEvidenceBytes`

The max evidence bytes should then vary by the difference in public key size and signature size, here 48.

The max vote size should vary by the signature size, here 32.

I wrote this a while ago so am not 100% sure of my logic. I guess I should have commented more.

File: types/block.go (hunk 9 above)
Comment (QuantumExplorer) [reply types/block.go:286]: Same reason as stated above.

File: state/tx_filter_test.go (hunk 8 above)
Comment (QuantumExplorer) [reply state/tx_filter_test.go:19]: There is a difference of 48 in the max evidence size and 32 in the max vote size so 80 total for 1 evidence and 1 validator.

File: crypto/crypto.go (hunk 3 above)
Comment (alexanderbez) [reply crypto/crypto.go:37]: That's OK. Typically `Type()` would return an enum, but this works too I suppose.

File: light/helpers_test.go (hunk 4 above)
Comment (alexanderbez) [reply light/helpers_test.go:49]: Yes :)

File: state/tx_filter.go (hunk 7 above)
Comment (alexanderbez) [reply state/tx_filter.go:14]: Makes sense! Yes, I'd add a small godoc on this explaining how/why.

**Review by alexanderbez — DISMISSED** [review]
Going to approve this 👍 

Also, any chance we can get CI tests to pass? I took a very brief look and it seems some of the C++/CGO stuff is missing?

File: CHANGELOG.md [hunk 10]
```diff
@@ -58,15 +58,15 @@ program](https://hackerone.com/tendermint).
 ### SECURITY:
```
Comment (shumkov) [inline CHANGELOG.md:58]: Let's just use a master version of the file?

File: abci/types/pubkey.go [hunk 11]
```diff
@@ -22,3 +24,17 @@ func Ed25519ValidatorUpdate(pk []byte, power int64) ValidatorUpdate {
 		Power:  power,
 	}
 }
+
+func BLS12381ValidatorUpdate(pk []byte, power int64) ValidatorUpdate {
+	pke := bls12381.PubKey(pk)
+	pkp, err := cryptoenc.PubKeyToProto(pke)
+	if err != nil {
+		panic(err)
```
Comment (shumkov) [inline abci/types/pubkey.go:32]: Is it ok to panic here? What could be a source of errors here?

File: consensus/replay_test.go [hunk 12]
```diff
@@ -657,16 +657,16 @@ func testHandshakeReplay(t *testing.T, config *cfg.Config, nBlocks int, mode uin
 	var commits []*types.Commit
```
Comment (shumkov) [inline consensus/replay_test.go:657]: Let's keep it as it is and make a separate PR on upstream?

File: consensus/wal.go [hunk 13]
```diff
@@ -202,7 +202,7 @@ func (wal *BaseWAL) WriteSync(msg WALMessage) error {
 	}
```
Comment (shumkov) [inline consensus/wal.go:202]: Let's keep it as it is?

File: docs/architecture/adr-024-sign-bytes.md [hunk 14]
```diff
@@ -2,28 +2,28 @@
```
Comment (shumkov) [inline docs/architecture/adr-024-sign-bytes.md:2]: Let's keep it as it is?

File: docs/architecture/adr-030-consensus-refactor.md [hunk 15]
```diff
@@ -85,7 +85,7 @@ func Consensus(Event, State) (State, Message) {
 ```
```
Comment (shumkov) [inline docs/architecture/adr-030-consensus-refactor.md:85]: Let's keep it as it is?

File: docs/architecture/adr-052-tendermint-mode.md [hunk 16]
```diff
@@ -7,7 +7,7 @@
```
Comment (shumkov) [inline docs/architecture/adr-052-tendermint-mode.md:7]: Let's keep it as it is?

File: libs/events/README.md [hunk 17]
```diff
@@ -26,7 +26,7 @@ Pub-Sub in go with event caching
```
Comment (shumkov) [inline libs/events/README.md:26]: Let's keep it as it is?

File: privval/socket_listeners.go [hunk 18]
```diff
@@ -1,10 +1,10 @@
 package privval
```
Comment (shumkov) [inline privval/socket_listeners.go:1]: Let's keep it as it is?

File: light/verifier_test.go (hunk 6 above)
Comment (QuantumExplorer) [reply light/verifier_test.go]: Fixed

File: light/helpers_test.go (hunk 4 above)
Comment (QuantumExplorer) [reply light/helpers_test.go]: Fixed

File: abci/types/pubkey.go (hunk 11 above)
Comment (QuantumExplorer) [reply abci/types/pubkey.go:32]: I copied the exact same logic as in Edwards keys.

Comment (shumkov) [reply abci/types/pubkey.go:32]: Discussed on call. It's good. The error could be possible only in case if we introduce a new key type and don't handle it in `PubKeyToProto` method.

**Comment by shumkov** [comment]
@QuantumExplorer hey! Why do you update markdown documents formatting in this PR?

**Review by shumkov — APPROVED** [review]
LGTM! Agreed to disabled some actions due to issue with cgo
